### PR TITLE
chore(scheduler-targets-alpha): imported function as a LambdaInvoke target results in synth-time error

### LINKED ISSUE #29284: (aws-scheduler-targets-alpha): Cannot use imported function ARN with `LambdaInvoke` target
### Describe the bug

I'm trying to invoke a "singleton" lambda created in another Stack. I expose the Function ARN via a CloudFormation output. If the lambda function ARN is being imported from a CFN output (e.g., `Fn.importValue`), even with `sameEnvironment: true`, an error is thrown at synth time.

### Expected Behavior

I expected the validation for the region to be ignored if I use `Lambda.fromFunctionAttributes` with the `sameEnvironment` value set to `true`.

### Current Behavior

I get this error:

```
/private/tmp/bug-report/node_modules/@aws-cdk/aws-scheduler-targets-alpha/lib/lambda-invoke.ts:21
      throw new Error(`Cannot assign function in region ${this.func.env.region} to the schedule ${Names.nodeUniqueId(schedule.node)} in region ${schedule.env.region}. Both the schedule and the function must be in the same region.`);
            ^
Error: Cannot assign function in region ${Token[TOKEN.22]} to the schedule BugReportStackSchedule10380292 in region us-east-2. Both the schedule and the function must be in the same region.
    at LambdaInvoke.addTargetActionToRole (/private/tmp/bug-report/node_modules/@aws-cdk/aws-scheduler-targets-alpha/lib/lambda-invoke.ts:21:13)
    at LambdaInvoke.bindBaseTargetConfig (/private/tmp/bug-report/node_modules/@aws-cdk/aws-scheduler-targets-alpha/lib/target.ts:80:10)
    at LambdaInvoke.bind (/private/tmp/bug-report/node_modules/@aws-cdk/aws-scheduler-targets-alpha/lib/target.ts:103:17)
    at new Schedule (/private/tmp/bug-report/node_modules/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:315:39)
    at new BugReportStack (/private/tmp/bug-report/lib/bug-report-stack.ts:29:5)
    at Object.<anonymous> (/private/tmp/bug-report/bin/bug-report.ts:7:1)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module.m._compile (/private/tmp/bug-report/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Object.require.extensions.<computed> [as .ts] (/private/tmp/bug-report/node_modules/ts-node/src/index.ts:1621:12)
```

### Reproduction Steps

```ts
import { Schedule, ScheduleExpression } from '@aws-cdk/aws-scheduler-alpha';
import { ScheduleTargetInput } from '@aws-cdk/aws-scheduler-alpha/lib/input';
import { LambdaInvoke } from '@aws-cdk/aws-scheduler-targets-alpha';
import * as cdk from 'aws-cdk-lib';
import { Fn } from 'aws-cdk-lib';
import { Function } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

export class BugReportStack extends cdk.Stack {
  constructor(scope: Construct, id: string) {
    super(scope, id, {
      env: {
        account: '123456789012',
        region: 'us-east-2',
      }
    });

    const functionArn = Fn.importValue("MyImportArn");
    const lambda = Function.fromFunctionAttributes(this, "ImportedFunction", {
      functionArn,
      sameEnvironment: true,
    });

    const target = new LambdaInvoke(lambda, {
      input: ScheduleTargetInput.fromObject({
        foo: "bar"
      }),
    });
    new Schedule(this, "Schedule", {
      schedule: ScheduleExpression.rate(cdk.Duration.minutes(5)),
      target,
    });
  }
}
```

### Possible Solution

If `sameEnvironment` is set to `true` on the imported lambda, the validations should be skipped.

### Additional Information/Context

Alpha versions are latest as well

```json
{
  "name": "bug-report",
  "version": "0.1.0",
  "bin": {
    "bug-report": "bin/bug-report.js"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "cdk": "cdk"
  },
  "devDependencies": {
    "@types/jest": "^29.5.12",
    "@types/node": "20.11.19",
    "aws-cdk": "2.130.0",
    "jest": "^29.7.0",
    "ts-jest": "^29.1.2",
    "ts-node": "^10.9.2",
    "typescript": "~5.3.3"
  },
  "dependencies": {
    "@aws-cdk/aws-scheduler-alpha": "^2.130.0-alpha.0",
    "@aws-cdk/aws-scheduler-targets-alpha": "^2.130.0-alpha.0",
    "aws-cdk-lib": "2.130.0",
    "constructs": "^10.0.0",
    "source-map-support": "^0.5.21"
  }
}
```

### CDK CLI Version

2.130.0 (build bd6e5ee)

### Framework Version

_No response_

### Node.js Version

20.x

### OS

MacOS

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by anan-k**
### Issue # (if applicable)

Closes #29284 

### Reason for this change

When a lambda function in another stack is imported by ARN and used as a `LambdaInvoke` target, a synth-time error occurred due to an issue with environment (account or region) validation `sameEnvDimension`. The validation compares two environment values such as account or region and returns true/false. The validation passes only if the values are the same (`SAME`) or both are unresolved (`BOTH_UNRESOLVED`), but since the region of the imported function was unresolved, it resulted in `ONE_UNRESOLVED`, causing the error. To resolve this, I modified the validation to allow the `ONE_UNRESOLVED` case.

The proposed solution in the issue was to skip validations if `sameEnvironment` is set to true on the imported lambda. However, I did not adopt this solution because the `sameEnvironment` attribute is not stored in the Lambda function itself and cannot be used to toggle the validation.

### Description of changes

I changed the `sameEnvDimension` function to allow the case where either one of the parameters is unresolved.
I made these changes based on the implementation in [aws-cdk-lib/aws-events-targets/lib/util.ts](https://github.com/anan-k/aws-cdk/blob/6d16337c9faed12716697a3b3af2a2be259b21b0/packages/aws-cdk-lib/aws-events-targets/lib/util.ts#L155)

### Description of how you validated changes

I added a unit test for the `LambdaInvoke` target to cover the case when an imported lambda function is used.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by anan-k** [comment]
Exemption Request: This PR just changes validation behavior and does not affect the integration test.  The changes are covered by unit tests.

**Review by aaythapa — COMMENTED** [review]
Hi, thank you for your contribution! Did a quick review and I'm not sure if this is the best way to fix this, see comment.

>The proposed solution in the issue was to skip validations if sameEnvironment is set to true on the imported lambda. However, I did not adopt this solution because the sameEnvironment attribute is not stored in the Lambda function itself and cannot be used to toggle the validation.

Agree, that doesn't work for now. Will do a deeper dive to see if there is a way to expose this property

**Comment by anan-k** [comment]
Hi Aayush, thank you for your review and comment.  Let me check it.

**Comment by anan-k** [comment]
@aaythapa 
I'm sorry for the late reply and thank you again for your comment.  

As you mentioned, this change is more of a fundamental solution to just resolve this issue rather than a minimum change. However, the proposed addition of the `sameEnvironment` property to Lambda is considered to have an even greater impact. Furthermore, due to the nature of the property, it may not be appropriate for Lambda to retain it.

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/util.ts [hunk 1]
```diff
@@ -5,7 +5,12 @@ import { Token, TokenComparison } from 'aws-cdk-lib';
...
  * are unresolved (in which case both are expted to be "current region" or "current account").
+ *
+ * Also returns true if one is unresolved (in which case we expect the unresolved dimension to match
+ * the resolved dimension, but it is up to the user to ensure this). Returning true here makes sure
+ * that we are not overly aggressive in producing a synth-time error.
+ *
  */
 export function sameEnvDimension(dim1: string, dim2: string) {
-  return [TokenComparison.SAME, TokenComparison.BOTH_UNRESOLVED].includes(Token.compareStrings(dim1, dim2));
+  return [TokenComparison.SAME, TokenComparison.ONE_UNRESOLVED, TokenComparison.BOTH_UNRESOLVED].includes(Token.compareStrings(dim1, dim2));
```
[the remark this answers is not in the document]
Comment (anan-k) [reply packages/@aws-cdk/aws-scheduler-targets-alpha/lib/util.ts:15]: See comment: https://github.com/aws/aws-cdk/pull/29615#issuecomment-2065488807

**Comment by paulhcsun** [comment]
Hi @anan-k, just pinging on this PR again to see if you're still interested in working on it. There's just the merge conflicts and comment by Kendra to be addressed.

**Comment by anan-k** [comment]
Hi @paulhcsun, thank you for pinging to me.  Based on the comments received, I will reconsider the impact on users when validation fails and the methods for correction.
However, please understand that the start of this work will be in two weeks.

**Comment by samson-keung** [comment]
I agree with @TheRealAmazonKendra that this validation is not worth while, even though it was good intention to let the customer know early that target in a different environment is not supported. Reasons:

1. At synth time, there are cases where CDK does not know about the actual environment (e.g. when the environment is using Tokens). In other words, the environment check works in some scenarios and not others. This creates more confusion than benefit.
2. IF cross env target support is added in the future, this check will become a blocker for CDK customers to set up cross env target.

@anan-k I will remove this check in all targets if you are not working on it already.

**Comment by samson-keung** [comment]
Closing this as a [new one](https://github.com/aws/aws-cdk/pull/31837) is opened for this.

**Pull request closed without merging**